src/locator.rs

```diff
@@ -0,0 +1,143 @@
+}
+
+struct AutolocateResult {
+    location: IPAddressInfo,
+    timestamp: Instant,
+}
+
+#[derive(Deserialize, Clone, Default)]
+pub struct IPAddressInfo {
+    // Required fields
+    pub ip: String,
+    pub latitude: f64,
+    pub longitude: f64,
+    pub city: String,
+
+    // Optional fields
+    pub version: Option<String>,
+    pub region: Option<String>,
+    pub region_code: Option<String>,
+    pub country: Option<String>,
+    pub country_name: Option<String>,
+    pub country_code: Option<String>,
+    pub country_code_iso3: Option<String>,
+    pub country_capital: Option<String>,
+    pub country_tld: Option<String>,
+    pub continent_code: Option<String>,
+    pub in_eu: Option<bool>,
+    pub postal: Option<String>,
+    pub timezone: Option<String>,
+    pub utc_offset: Option<String>,
+    pub country_calling_code: Option<String>,
+    pub currency: Option<String>,
+    pub currency_name: Option<String>,
+    pub languages: Option<String>,
+    pub country_area: Option<f64>,
+    pub country_population: Option<f64>,
+    pub asn: Option<String>,
+    pub org: Option<String>,
+}
+
+#[derive(Deserialize, Debug, Default, Clone)]
+#[serde(default, deny_unknown_fields)]
+pub struct Locator {
+    pub driver: LocatorDriver,
+    #[serde(default = "getenv_api_key")]
+    api_key: Option<String>,
+}
+
+fn getenv_api_key() -> Option<String> {
+    std::env::var(API_KEY_ENV).ok()
+}
+
+#[derive(Deserialize, Debug, SmartDefault, Clone)]
+#[serde(rename_all = "lowercase", deny_unknown_fields)]
+pub enum LocatorDriver {
+    #[default]
+    Ipapi,
+    Ip2Location,
+}
+
+#[async_trait]
+pub trait Backend {
+    fn name(&self) -> Cow<'static, str>;
+
+    async fn get_info(
+        &self,
+        client: &reqwest::Client,
+        api_key: &Option<String>,
+    ) -> Result<IPAddressInfo>;
+}
+
+#[async_trait]
+impl Backend for LocatorDriver {
+    fn name(&self) -> Cow<'static, str> {
+        match self {
+            LocatorDriver::Ipapi => ipapi::Ipapi.name(),
+            LocatorDriver::Ip2Location => ip2location::Ip2Location.name(),
+        }
+    }
+
+    async fn get_info(
+        &self,
+        client: &reqwest::Client,
+        api_key: &Option<String>,
+    ) -> Result<IPAddressInfo> {
+        match self {
+            LocatorDriver::Ipapi => ipapi::Ipapi.get_info(client, api_key).await,
+            LocatorDriver::Ip2Location => ip2location::Ip2Location.get_info(client, api_key).await,
+        }
+    }
+}
+
+/// No-op if last API call was made in the last `interval` seconds.
+pub async fn find_ip_location(
+    client: &reqwest::Client,
+    interval: Duration,
+) -> Result<IPAddressInfo> {
+    {
+        let guard = LAST_AUTOLOCATE.lock().unwrap();
+        if let Some(cached) = &*guard {
+            if cached.timestamp.elapsed() < interval {
+                return Ok(cached.location.clone());
+            }
+        }
+    }
+
+    let locator = LOCATOR.lock().unwrap().clone();
```

Comment:
perhaps another kind of mutex/rwlock could be used across the await so that it doesn't have to be cloned...

Reply:
If we store `Locator::api_key` as an `Arc<str>`, this clone will be cheap.

Reply:
Wouldn't it need to be `Arc<Option<String>>`?

Reply:
or `Option<Arc<str>>`. Or `Locator` can be wrapped in `Arc`.